org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/batch/ClasspathJsr199.java

```diff
@@ -195,6 +216,29 @@ public char[][] getModulesDeclaringPackage(String aQualifiedPackageName, String
 		return singletonModuleNameIf(result);
 	}
 
+	@Override
+	public char[][] listPackages() {
+		Set<String> packageNames = new HashSet<>();
+		try {
+			for (JavaFileObject fileObject : this.fileManager.list(this.location, "", fileTypes, true)) { //$NON-NLS-1$
```

Comment:
Listing all class files in all packages is surely a performance issue. Also passing "" to signal no filtering is a bit crude (needed an adjustment inside ModuleFinder, too). Unfortunately I found no better we to list all packages, which, however, the compiler needs particularly for auto modules.